presto-main/src/test/java/com/facebook/presto/operator/scalar/StringFunctionsBenchmark.java

```diff
@@ -0,0 +1,274 @@
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static org.openjdk.jmh.annotations.Mode.AverageTime;
+import static org.openjdk.jmh.annotations.Scope.Thread;
+
+@SuppressWarnings("MethodMayBeStatic")
+@State(Thread)
+@OutputTimeUnit(NANOSECONDS)
+@BenchmarkMode(AverageTime)
+@Fork(1)
+@Warmup(iterations = 4, time = 500, timeUnit = MILLISECONDS)
+@Measurement(iterations = 5, time = 500, timeUnit = MILLISECONDS)
+public class StringFunctionsBenchmark
+{
+    @Benchmark
+    public long benchmarkLength(BenchmarkData data)
+    {
+        return length(data.getSlice());
+    }
+
+    @Benchmark
+    public Slice benchmarkSubstringStart(BenchmarkData data)
+    {
+        Slice slice = data.getSlice();
+        int length = data.getLength();
+        return substr(slice, (length / 2) - 1);
+    }
+
+    @Benchmark
+    public Slice benchmarkSubstringStartLength(BenchmarkData data)
+    {
+        Slice slice = data.getSlice();
+        int length = data.getLength();
+        return substr(slice, (length / 2) - 1, length / 2);
+    }
+
+    @Benchmark
+    public Slice benchmarkSubstringStartFromEnd(BenchmarkData data)
+    {
+        Slice slice = data.getSlice();
+        int length = data.getLength();
+        return substr(slice, -((length / 2) + 1));
+    }
+
+    @Benchmark
+    public Slice benchmarkSubstringStartLengthFromEnd(BenchmarkData data)
+    {
+        Slice slice = data.getSlice();
+        int length = data.getLength();
+        return substr(slice, -((length / 2) + 1), length / 2);
+    }
+
+    @Benchmark
+    public Slice benchmarkReverse(BenchmarkData data)
+    {
+        return reverse(data.getSlice());
+    }
+
+    @Benchmark
+    public Slice benchmarkLeftTrim(WhitespaceData data)
+    {
+        return leftTrim(data.getLeftWhitespace());
+    }
+
+    @Benchmark
+    public Slice benchmarkRightTrim(WhitespaceData data)
+    {
+        return rightTrim(data.getRightWhitespace());
+    }
+
+    @Benchmark
+    public Slice benchmarkTrim(WhitespaceData data)
+    {
+        return trim(data.getBothWhitespace());
+    }
+
+    @Benchmark
+    public Slice benchmarkUpper(WhitespaceData data)
+    {
+        return upper(data.getBothWhitespace());
+    }
+
+    @Benchmark
+    public Slice benchmarkLower(WhitespaceData data)
+    {
+        return lower(data.getBothWhitespace());
+    }
+
+    @State(Thread)
+    public static class BenchmarkData
+    {
+        private static final int[] ASCII_CODE_POINTS;
+        private static final int[] ALL_CODE_POINTS;
+
+        static {
+            ASCII_CODE_POINTS = new int[0x7F];
+            for (int i = 0; i < ASCII_CODE_POINTS.length; i++) {
+                ASCII_CODE_POINTS[i] = i;
+            }
+
+            int[] codePoints = new int[MAX_CODE_POINT];
+            int index = 0;
+            for (int codePoint = 0; codePoint < MAX_CODE_POINT; codePoint++) {
+                if (Character.isDefined(codePoint)) {
```

Comment:
Looks like this code is not up to date. As is return, surrogate characters will be included.